src/Aspire.Hosting/EventingExtensions.cs

```diff
@@ -64,13 +64,13 @@ public static IResourceBuilder<T> OnResourceEndpointsAllocated<T>(this IResource
     /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
     public static IResourceBuilder<T> OnResourceReady<T>(this IResourceBuilder<T> builder, Func<T, ResourceReadyEvent, CancellationToken, Task> callback)
         where T : IResource
-        => builder.OnEvent(callback);
+        => builder.OnEvent(builder.Resource, callback);
 
-    private static IResourceBuilder<TResource> OnEvent<TResource, TEvent>(this IResourceBuilder<TResource> builder, Func<TResource, TEvent, CancellationToken, Task> callback)
+    private static IResourceBuilder<TResource> OnEvent<TResource, TArg, TEvent>(this IResourceBuilder<TResource> builder, TArg callbackArgument, Func<TArg, TEvent, CancellationToken, Task> callback)
```

Comment:
I think that new signature is a bit hard to parse.

I get why we are doing it, but I think when using the On*Event(...) methods tied to a resource we either support getting the resource builder, or the event, but not both.

We possibly also want to mark these APIs as experimental for 9.4 just until we've had enough time to flush out any potential issues with them.

What you'd probably do is inside OnEvent is construct a callback that captures the resource builder and passes it into the callback which is triggered by the underling Eventing callback and then pass along the captured resource builder into the callback passed into the On* methods.

Whilst the TArg trick is kind of cute when I looked at the signature it took me a moment to figure out why it was even there 🙂

Reply:
If we really want to support the resource and resource builder two overloads might make more sense unless it created ambiguity somehow.

Reply:
> I get why we are doing it, but I think when using the On*Event(...) methods tied to a resource we either support getting the resource builder, or the event, but not both.

If you had a resource builder but not the event, you wouldn't have the Service Provider, which feels like a bit of a miss.

> We possibly also want to mark these APIs as experimental for 9.4 just until we've had enough time to flush out any potential issues with them.

I was also thinking last night about how this would interact with #9042, as well as possibly any equivalent for non resource events on `IDistributedApplicationBuilder`, so I think it's more than fair to mark these as experimental.

Reply:
Another idea, what if we make the underlying event object support a typed resource, then we don't need the separate resource param.  Can also then include the resource builder itself on appropriate events

```
class ResourceReadyEvent<T>(IResourceBuilder<T> builder, IServiceProvider services)
   : ResourceReadyEvent(builder.Resource, services)
   where T : IResource
{
   public new T Resource => builder.Resource;

   public IResourceBuilder<T> Builder => builder;
}
```

This could also be retrofitted in a backwards compatible way with the existing event subscriptions (May need to sort out #5565 ) though

Reply:
I've split the `OnEvent` method to avoid the `TArg` trick.  

Any process for defining the `diagnosticId` to mark the methods as experimetnal.  `ASPIREEVENTING001` would seem to follow the existing conventions.

@mitchdenny Any thoughts on how to deal with the Service Provider if you had the resource builder without the event object?

Reply:
If you want to keep all events with `IResource`, then there's nothing to do - that's already in main.  I can back out the refactor to `OnBeforeResourceStart` to pass `IResourceBuilder<T>` and then we can merge this for 9.5 - it only transfers some uses over to the new APIs, so no need to squeeze it in to 9.4.

If we do want to use IResourceBuilder<T> for `OnBeforeResoruceStart` that is in this PR.

I also have https://github.com/afscrome/aspire/commit/cde88b1813496cf83cabe3c8f2bbb9dcc061aad0 for using `IResourceBuilder<T>` with `ResourceInitialized()`.  I thought this was an event that would benefit from having access to the resource builder, however none of the 3 implementations in Aspire made use of it so perhaps it's not as useful as I first thought.